### ISSUE TITLE
 how to infer image of  a segmentation or depth like the demo 

### ISSUE THREAD

**Issue by angrysword** [facebookresearch/dinov2#6]
have not find an easy way to do that. managed to get features. 

`import torch
from PIL import Image
import torchvision.transforms as T
import hubconf

dinov2_vits14 = hubconf.dinov2_vits14()

image_transforms = T.Compose([
    T.Resize(256, interpolation=T.InterpolationMode.BICUBIC),
    T.CenterCrop(224),
    T.ToTensor(),
    T.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
])

img = Image.open('pic.jpg')
img = image_transforms(img)[:3].unsqueeze(0)

features=dinov2_vits14.forward_features(img)

for k, v in features.items():
    if v is not None:
        print(k, v.shape)
    else:
        print(k, v)
`

x_norm_clstoken torch.Size([1, 384])
x_norm_patchtokens torch.Size([1, 256, 384])
x_prenorm torch.Size([1, 257, 384])
masks None

**Comment by SikanderBinMukaram** [comment]
I was also able to get features but do you have any idea how we can extract the depth info from the features patch tokens?

**Comment by woctezuma** [comment]
![Paper](https://i.imgur.com/Xj2CZro.png)

**Comment by patricklabatut** [comment]
Thanks @woctezuma for helping out and sharing the relevant pointers in our paper.

For segmentation and depth estimation,

**Comment by dsnsabari** [comment]
@patricklabatut, are you planning to release the depth estimation model with the test code?

**Comment by GaussRieman** [comment]
Looking forward to it too!

**Comment by woctezuma** [comment]
> @patricklabatut, are you planning to release the depth estimation model with the test code?

- https://github.com/facebookresearch/dinov2/issues/14

**Comment by angrysword** [comment]
Thanks so much. Hopefully release the head as well as pretrained model. It is very hard to understand what document saying, I have read a couple of times. 
Question 1. the last layer have these. x_norm_clstoken torch.Size([1, 384])  x_norm_patchtokens torch.Size([1, 256, 384]). definitely we can concate it together. but base on which dimension
Question 2. we upsample it by 4. then use linear, and out put 256. that is a lot parameters 256*384*4*256 there, will compute cost is too much?
Question 3. why we talk about dpt decoder, if linear works well.

code is much better than the document explain.

**Comment by stellanhaglund** [comment]
> Thanks @woctezuma for helping out and sharing the relevant bits from our paper.
> 
> For segmentation and depth estimation, training a linear head or a more complex one is needed to achieve results similar to the ones shown in the paper or the demo. We will consider including more pointers or helpers to facilitate reproducing these.

It's the same for object classification with bounding-boxes then?

**Comment by JuliusJacobsohn** [comment]
I'm also looking forward to the depth estimation features.

**Comment by patricklabatut** [comment]
Closing as answered and will be using #54 and #55 to keep track of requests around documentation, training code or model weights for depth or segmentation. Thanks for the feedback!

**Comment by Cindy0725** [comment]
> 

Hi @angrysword, thank you for opening this issue. Have you figured out these questions successfully? I am confused by the dimension of the tensors. How can we convert tensor ([batch_size, 256, 384]) to a depth tensor with size ([batch_size, 256, image size, image_size])? We can't achieve this by only doing bilinear upsampling.

**Comment by patricklabatut** [comment]
We have just added the models and supporting code for semantic segmentation and depth estimation.
